File: android/src/main/kotlin/com/tophap/mapbox_gl/MapboxMapFactory.kt
package com.tophap.mapbox_gl

import android.content.Context
import com.tophap.mapbox_gl.proto.Mapbox
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class MapboxMapFactory(private val registrar: PluginRegistry.Registrar) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val proto = Mapbox.Map.Options.parseFrom(args as ByteArray)

        val platformViewChannel = MethodChannel(registrar.messenger(), "com.tophap/mapbox_gl_factory_$viewId")
        return MapboxPlatformView(context.applicationContext, proto, platformViewChannel, viewId)
    }
}

File: ios/Classes/MapboxMapFactory.swift
//
// Created by Razvan Lung on 2019-08-08.
//

import Foundation
import Flutter

class MapboxMapFactory: NSObject, FlutterPlatformViewFactory {
  let registrar: FlutterPluginRegistrar

  init(registrar: FlutterPluginRegistrar) {
    self.registrar = registrar
  }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    let data = (args as! FlutterStandardTypedData).data
    let proto = try! Tophap_MapboxGl_Map.Options(serializedData: data)

    let platformViewChannel = FlutterMethodChannel(name: "com.tophap/mapbox_gl_factory_\(viewId)", binaryMessenger: registrar.messenger())
    return MapboxPlatformView(withFrame: frame, options: proto, channel: platformViewChannel, viewId: viewId)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}
